api/oc_push.c

```diff
@@ -0,0 +1,2605 @@
+					}
+
+				}
+
+				oc_free_string(&recv_obj->receiveruri);
+				oc_new_string(&recv_obj->receiveruri, oc_string(rep->value.string), oc_string_len(rep->value.string));
+			}
+			break;
+
+		case OC_REP_STRING_ARRAY:
+			if (!strcmp(oc_string(rep->name), "rts"))
+			{
+				oc_free_string_array(&recv_obj->rts);
+				int len = oc_string_array_get_allocated_size(rep->value.array);
+				oc_new_string_array(&recv_obj->rts, len);
+
+				for (int i=0; i < len; i++)
+				{
+					oc_string_array_add_item(recv_obj->rts, oc_string_array_get_item(rep->value.array, i));
+				}
+			}
+			break;
+
+		default:
+			OC_PUSH_ERR("something wrong, unexpected Property type: %d", rep->type);
+			return;
+		}
+		rep = rep->next;
+	}
+
+	return;
+}
+
+
+
+/*
+ * @brief				create new receiver object
+ *
+ * @param recv_obj	pointer for new receiver object
+ * @param rep			received new receiver object object
+ */
+static bool _create_recv_obj(oc_recvs_t *recvs_instance, oc_rep_t *rep)
+{
+	bool result = false;
+	char mandatory_property_check = 0;
+	oc_recv_t *recv_obj = (oc_recv_t *)oc_memb_alloc(&recv_instance_memb);
+
+	if (!recv_obj)
+	{
+		OC_PUSH_ERR("oc_memb_alloc() error!");
+		return result;
+	}
+
+	while (rep)
+	{
+		switch (rep->type)
+		{
+		case OC_REP_STRING:
+			if (!strcmp(oc_string(rep->name), "receiveruri"))
+			{
+				oc_new_string(&recv_obj->receiveruri, oc_string(rep->value.string), oc_string_len(rep->value.string));
+				mandatory_property_check |= 0x1;
+			}
+			break;
+
+		case OC_REP_STRING_ARRAY:
+			if (!strcmp(oc_string(rep->name), "rts"))
+			{
+				int len = oc_string_array_get_allocated_size(rep->value.array);
+				oc_new_string_array(&recv_obj->rts, len);
+
+				for (int i=0; i < len; i++)
+				{
+					oc_string_array_add_item(recv_obj->rts, oc_string_array_get_item(rep->value.array, i));
+				}
+
+				mandatory_property_check |= 0x2;
+			}
+			break;
+
+		default:
+			OC_PUSH_ERR("something wrong, unexpected Property type: %d", rep->type);
+			break;
+		}
+		rep = rep->next;
+	}
+
+	if (mandatory_property_check != 0x3)
+		return result;
+	else
+		result = true;
+
+	oc_list_add(recvs_instance->receivers, recv_obj);
+
+	/* create app resource corresponding to receiver object */
+	OC_PUSH_DBG("new app resource for new receiver obj (\"%s\") is created...", oc_string(recv_obj->receiveruri));
+	if (_create_pushd_rsc(recv_obj, recvs_instance->resource))
+		result = true;
+
+	return result;
+}
+
+
+
+
+/*
+ * @brief					replace existing receiver object array with new one
+ *
+ * @param recv_obj_list	receiver object array
+ * @param resource		Push Receiver resource
+ * @param rep				payload representation of new receiver object array
+ */
+static bool _replace_recv_obj_array(oc_recvs_t *recvs_instance, oc_rep_t *rep)
+{
+	int obj_arr_len;
+	oc_rep_t *rep_obj;
+	bool result = true;
+
+	/* remove existing receivers object array */
+	_purge_recv_obj_list(recvs_instance);
```

Comment:
Wouldn't it be better to keep the recv_obj_list intact in case some invocation `_create_recv_obj` fails.

Usually the stuff is done in the following fashion:
1) prepare new data to some temporarily list (in this case this would be the `while` cycle, but it would write the data to a new oc_list)
2) if an error occurs deallocate stuff in the temp list, write error and return
3) if everything is in order -> deallocate the original list and replace it with the new list

Though this would potentially have higher memory requirements, because for a moment you'd hold both the old and new lists in memory. This shouldn't be a problem if the lists aren't big.

Reply:
Ok, I did like this:
1. validate received receiver object list first
2. if validation fails just return false (original data is intact), if validation is success do remaining job

Reply:
Yeah, that works too, good job!